File: python/euler/bin/p0022.py
# project euler: problem 22

from collections.abc import Iterable
from typing import IO


def total_score(words: Iterable[str]) -> int:
    def score(word):
        return sum(ord(c) - (ord('A') - 1) for c in word)

    return sum(idx * score(word) for idx, word in enumerate(words, start=1))


def compute(fh: IO) -> str:
    keywords = [s.strip('"') for s in fh.read().split(',')]

    return str(total_score(sorted(keywords)))


def solve() -> str:
    from euler.lib.resource import asset_file

    fh = asset_file('p022_names.txt')
    result = compute(fh)
    fh.close()
    return result

File: typescript/euler/bin/p0022.ts
// project euler: problem 22

import { assetData } from "../lib/asset.ts";
import { sum } from "../lib/math.ts";

const calcScore = (words: string[]): number => {
  const score = (word: string): number => {
    return sum(
      word.split("").map((x) => x.charCodeAt(0) - "A".charCodeAt(0) + 1),
    );
  };

  let acc = 0;
  for (const [idx, word] of words.entries()) {
    acc += (idx + 1) * score(word);
  }

  return acc;
};

export const compute = (data: string): string => {
  const keywords = data.replaceAll('"', "").split(",");
  keywords.sort();

  return String(calcScore(keywords));
};

export const solve = (): string => {
  const data = new TextDecoder().decode(assetData("p022_names.txt"));
  return compute(data);
};
